### ISSUE TITLE
Wrong $" warning in perl 5.14

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11750]
Migrated from [rt.perl.org#103766](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=103766) (status was 'resolved')

Searchable as RT103766$

**Comment by p5pRT** [comment]
### From am@jobindex.dk

This warning in perl 5\.14 had us confused for a while&#8203;:

$ perl \-wE 'say $^V; my $x; say "@&#8203;\{\[ $x \]\}";'
v5\.14\.2
Use of uninitialized value $" in join or string at \-e line 1\.

It blames $"\, though it is in fact the variable $x that is 
uninitialized\. It used to be correct&#8203;:

$ perl \-wE 'say $^V; my $x; say "@&#8203;\{\[ $x \]\}";'
v5\.12\.4
Use of uninitialized value in join or string at \-e line 1\.


Interestingly\, while the 5\.14 change appears to be wrong\, it is actually 
helpful in that it points towards the "@&#8203;array" construction\, even if 
misses the target by a little bit\.


\(Setting $" to undef will give an additional and correct warning in both 
cases\)


Cheers\,
Anders\.

**Comment by p5pRT** [comment]
### From @pjcj

On Wed\, Nov 16\, 2011 at 04&#8203;:57&#8203;:21AM \-0800\, Anders Melchiorsen wrote&#8203;:
> \# New Ticket Created by  Anders Melchiorsen 
> \# Please include the string&#8203;:  \[perl \#103766\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=103766 >
> 
> 
> This warning in perl 5\.14 had us confused for a while&#8203;:
> 
> $ perl \-wE 'say $^V; my $x; say "@&#8203;\{\[ $x \]\}";'
> v5\.14\.2
> Use of uninitialized value $" in join or string at \-e line 1\.
> 
> It blames $"\, though it is in fact the variable $x that is 
> uninitialized\. It used to be correct&#8203;:
> 
> $ perl \-wE 'say $^V; my $x; say "@&#8203;\{\[ $x \]\}";'
> v5\.12\.4
> Use of uninitialized value in join or string at \-e line 1\.
> 
> 
> Interestingly\, while the 5\.14 change appears to be wrong\, it is actually 
> helpful in that it points towards the "@&#8203;array" construction\, even if 
> misses the target by a little bit\.
> 
> 
> \(Setting $" to undef will give an additional and correct warning in both 
> cases\)
> 
> 
> Cheers\,
> Anders\.

$ \.\./perl\-git/Porting/bisect\.pl \-j 6 \-Dnoextensions=Encode \-e 'use warnings "uninitialized"; $SIG\{\_\_WARN\_\_\} = sub \{ die if "@&#8203;\_" =~ /\\$"/ \}; "@&#8203;\{\[$x\]\}"'

This was caused by&#8203;:

commit 6d1f0892ce0bd77f843552ab189aa5f121c374d4
Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;:   Fri Dec 10 14&#8203;:54&#8203;:13 2010 \-0800

&nbsp;    \[perl \#72090\] unitialized variable name wrong with no strict refs
&nbsp;    
&nbsp;    $ \./perl \-we '$a = @&#8203;$a > 0'
&nbsp;    Use of uninitialized value $a in array dereference at \-e line 1\.
&nbsp;    Use of uninitialized value $a in numeric gt \(>\) at \-e line 1\.
&nbsp;    
&nbsp;    S\_find\_uninit\_var was not taking into account that rv2\*v could return
&nbsp;    undef\. So it merrily looked at the child ops to find one that named
&nbsp;    a variable\.
&nbsp;    
&nbsp;    This commit makes it skip any rv2av/rv2hv that does not have an OP\_GV
&nbsp;    as its child op\.
&nbsp;    
&nbsp;    In other words\, it skips @&#8203;\{\.\.\.\} and %\{\.\.\.\} \(including the shorthand
&nbsp;    forms @&#8203;$foo and %$foo\)\, but not @&#8203;foo or %foo\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/sv.c b/sv.c
index c0c2458..2cabf7b 100644
--- a/sv.c
+++ b/sv.c
@@ -14031,6 +14031,12 @@ S_find_uninit_var(pTHX_ const OP *const obase, const SV *const uninit_sv,
                if ( (type == OP_CONST && SvOK(cSVOPx_sv(kid)))
                  || (type == OP_NULL  && ! (kid->op_flags & OPf_KIDS))
                  || (type == OP_PUSHMARK)
+                 || (
+                     /* @$a and %$a, but not @a or %a */
+                       (type == OP_RV2AV || type == OP_RV2HV)
+                    && cUNOPx(kid)->op_first
+                    && cUNOPx(kid)->op_first->op_type != OP_GV
+                    )
                )
                continue;
            }
diff --git a/t/lib/warnings/sv b/t/lib/warnings/sv
index dbab90b..e29553a 100644
--- a/t/lib/warnings/sv
+++ b/t/lib/warnings/sv
@@ -209,6 +209,13 @@ Use of uninitialized value $a in join or string at - line 4.
 Use of uninitialized value $a in concatenation (.) or string at - line 5.
 Use of uninitialized value $a in concatenation (.) or string at - line 6.
 ########
+# [perl #72090]
+use warnings 'uninitialized';
+$a = @$a > 0;
+EXPECT
+Use of uninitialized value $a in array dereference at - line 3.
+Use of uninitialized value in numeric gt (>) at - line 3.
+########
 # sv.c 
 use warnings 'numeric' ;
 sub TIESCALAR{bless[]} ; 



```
</details>
And, indeed, reverting the patch fixes the bug and reinstates the
previous one\.

Here's a patch to add a test for it \(also attached\)&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/t/lib/warnings/sv b/t/lib/warnings/sv
index ae65e51..86b50e0 100644
--- a/t/lib/warnings/sv
+++ b/t/lib/warnings/sv
@@ -216,6 +216,12 @@ EXPECT
 Use of uninitialized value $a in array dereference at - line 3.
 Use of uninitialized value in numeric gt (>) at - line 3.
 ########
+# [perl #103766]
+use warnings 'uninitialized';
+"@{[$x]}";
+EXPECT
+Use of uninitialized value in join or string at - line 3.
+########
 # sv.c 
 use warnings 'numeric' ;
 sub TIESCALAR{bless[]} ; 



```
</details>
It would be even nicer $x could be identified in the error message.

\-\- 
Paul Johnson \- paul@&#8203;pjcj\.net
http&#8203;://www\.pjcj\.net

**Comment by p5pRT** [comment]
### From @pjcj

<details><summary>Inline Patch</summary>

```diff
diff --git a/t/lib/warnings/sv b/t/lib/warnings/sv
index ae65e51..86b50e0 100644
--- a/t/lib/warnings/sv
+++ b/t/lib/warnings/sv
@@ -216,6 +216,12 @@ EXPECT
 Use of uninitialized value $a in array dereference at - line 3.
 Use of uninitialized value in numeric gt (>) at - line 3.
 ########
+# [perl #103766]
+use warnings 'uninitialized';
+"@{[$x]}";
+EXPECT
+Use of uninitialized value in join or string at - line 3.
+########
 # sv.c 
 use warnings 'numeric' ;
 sub TIESCALAR{bless[]} ; 
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Nov 16 17&#8203;:18&#8203;:43 2011\, paul@&#8203;pjcj\.net wrote&#8203;:
> It would be even nicer $x could be identified in the error message\.
> 

I don’t think that’s possible\, as the uninitialised variable is the
first element of the array\, which is a copy of $x\, not $x itself\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Nov 18 18&#8203;:04&#8203;:23 2011\, sprout wrote&#8203;:
> On Wed Nov 16 17&#8203;:18&#8203;:43 2011\, paul@&#8203;pjcj\.net wrote&#8203;:
> > It would be even nicer $x could be identified in the error message\.
> > 
> 
> I don’t think that’s possible\, as the uninitialised variable is the
> first element of the array\, which is a copy of $x\, not $x itself\.

Actually\, it is possible\.  But you are asking for another variation of
the same bug \(\#72090\) that I fixed with 6d1f0892c\.  And now that I
understand the code better\, I’m sure whether that was a bug that needed
fixing\.  It seems the code was designed to give not only the name of the
variable itself that was undefined\, but the variable that the undefined
value was derived from\, if possible\.  That would make bug \#72090 not a
bug\, and therefore the fix should simply be reverted\.  But I’m not sure\.

Bug \#72090 was that

&nbsp;    $a = @&#8203;$a > 0

mentions $a in the warning message\.  @&#8203;\{\.\.\.\}\, outside of strict refs\, can
itself return undef if its argument is undef\.  Hence the undefined value
is ultimately derived from\, but not the same scalar as\, $a\.

Whether that is considered a bug will\, of course\, affect how I fix
\*this\* bug\.

I see that Dave Mitchell added this feature to begin with in commit
29489e7c7\, so\, Dave\, if you are reading this\, would you please chime in?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Nov 18\, 2011 at 08&#8203;:24&#8203;:07PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> On Fri Nov 18 18&#8203;:04&#8203;:23 2011\, sprout wrote&#8203;:
> > On Wed Nov 16 17&#8203;:18&#8203;:43 2011\, paul@&#8203;pjcj\.net wrote&#8203;:
> > > It would be even nicer $x could be identified in the error message\.
> > > 
> > 
> > I don’t think that’s possible\, as the uninitialised variable is the
> > first element of the array\, which is a copy of $x\, not $x itself\.
> 
> Actually\, it is possible\.  But you are asking for another variation of
> the same bug \(\#72090\) that I fixed with 6d1f0892c\.  And now that I
> understand the code better\, I’m sure whether that was a bug that needed
> fixing\.  It seems the code was designed to give not only the name of the
> variable itself that was undefined\, but the variable that the undefined
> value was derived from\, if possible\.  That would make bug \#72090 not a
> bug\, and therefore the fix should simply be reverted\.  But I’m not sure\.
> 
> Bug \#72090 was that
> 
>     $a = @&#8203;$a > 0
> 
> mentions $a in the warning message\.  @&#8203;\{\.\.\.\}\, outside of strict refs\, can
> itself return undef if its argument is undef\.  Hence the undefined value
> is ultimately derived from\, but not the same scalar as\, $a\.
> 
> Whether that is considered a bug will\, of course\, affect how I fix
> \*this\* bug\.
> 
> I see that Dave Mitchell added this feature to begin with in commit
> 29489e7c7\, so\, Dave\, if you are reading this\, would you please chime in?

Well\, nothing's entirely clear cut\.

However\, suppose hypothetically that sin\(undef\) returned undef \(rather
than 0\.0\); then I would expect the following&#8203;:

&nbsp;    $ perl \-we 'my \($x\,$y\); $y = 1 \+ sin $x'
&nbsp;    Use of uninitialized value $x in sin at \-e line 1\.
&nbsp;    Use of uninitialized value in addition \(\+\) at \-e line 1\.
&nbsp;    $

i\.e\. the second warning \*shouldn't\* mention $x\.
So I think your fix for \#72090 was correct\.

As for the OP's case\,  say @&#8203;\{\[ $x \]\}&#8203;: clearly the current behaviour of
blaming $" is wrong\, and certainly fixing so that it blames no variable is
correct\. Whether it should be extra smart and blame $x is debatable\, but
on balance I think not\. When looking at the SV that triggered the
warninng\, my code distinguishes between a specific SV that happens to be
\!SvOK\(\)\, and PL\_sv\_undef\. In the former case it can be quite aggressive in
matching that SV up with a particular pad or package variable\, while in
the latter case\, it has to be very conservative\, and usually only follow
certain type of op back\. @&#8203;\{\[$x\]\} is a special case where the undef value
unambiguously came from $x; but it fails in the more general case&#8203;:

&nbsp;    say @&#8203;\{\[$x\, $y\]\}

here\, using the elements of the array will raise two 'uninit' warnings\,
each of which may have been derived from either $x or $y\.


\-\- 
Nothing ventured\, nothing lost\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Nov 21 09&#8203;:25&#8203;:42 2011\, davem wrote&#8203;:
> On Fri\, Nov 18\, 2011 at 08&#8203;:24&#8203;:07PM \-0800\, Father Chrysostomos via RT
wrote&#8203;:
> > On Fri Nov 18 18&#8203;:04&#8203;:23 2011\, sprout wrote&#8203;:
> > > On Wed Nov 16 17&#8203;:18&#8203;:43 2011\, paul@&#8203;pjcj\.net wrote&#8203;:
> > > > It would be even nicer $x could be identified in the error message\.
> > > > 
> > > 
> > > I don’t think that’s possible\, as the uninitialised variable is the
> > > first element of the array\, which is a copy of $x\, not $x itself\.
> > 
> > Actually\, it is possible\.  But you are asking for another variation of
> > the same bug \(\#72090\) that I fixed with 6d1f0892c\.  And now that I
> > understand the code better\, I’m sure whether that was a bug that needed
> > fixing\.  It seems the code was designed to give not only the name of the
> > variable itself that was undefined\, but the variable that the undefined
> > value was derived from\, if possible\.  That would make bug \#72090 not a
> > bug\, and therefore the fix should simply be reverted\.  But I’m not sure\.
> > 
> > Bug \#72090 was that
> > 
> >     $a = @&#8203;$a > 0
> > 
> > mentions $a in the warning message\.  @&#8203;\{\.\.\.\}\, outside of strict refs\, can
> > itself return undef if its argument is undef\.  Hence the undefined value
> > is ultimately derived from\, but not the same scalar as\, $a\.
> > 
> > Whether that is considered a bug will\, of course\, affect how I fix
> > \*this\* bug\.
> > 
> > I see that Dave Mitchell added this feature to begin with in commit
> > 29489e7c7\, so\, Dave\, if you are reading this\, would you please chime in?
> 
> Well\, nothing's entirely clear cut\.
> 
> However\, suppose hypothetically that sin\(undef\) returned undef \(rather
> than 0\.0\); then I would expect the following&#8203;:
> 
>     $ perl \-we 'my \($x\,$y\); $y = 1 \+ sin $x'
>     Use of uninitialized value $x in sin at \-e line 1\.
>     Use of uninitialized value in addition \(\+\) at \-e line 1\.
>     $
> 
> i\.e\. the second warning \*shouldn't\* mention $x\.
> So I think your fix for \#72090 was correct\.
> 
> As for the OP's case\,  say @&#8203;\{\[ $x \]\}&#8203;: clearly the current behaviour of
> blaming $" is wrong\, and certainly fixing so that it blames no variable is
> correct\. Whether it should be extra smart and blame $x is debatable\, but
> on balance I think not\. When looking at the SV that triggered the
> warninng\, my code distinguishes between a specific SV that happens to be
> \!SvOK\(\)\, and PL\_sv\_undef\. In the former case it can be quite aggressive in
> matching that SV up with a particular pad or package variable\, while in
> the latter case\, it has to be very conservative\, and usually only follow
> certain type of op back\.

One of us must be confused\, as\, at the top level \(the first call\)\,
find\_uninit\_var does not make such a distinction based on the variable
itself\, but on the op type\. Only if the op is one that can return undef
for a \*defined\* value\, or if the op has multiple children that are
possible suspects\, does it look for an exact match\.  If an op has one
child\, it assumes it can simply dig deeper\.  That’s why I raised the
question as to whether \#72090 was really a bug\, because it appeared to
be by design\.

Based on your ‘I think not’\, why not change the code always to check for
an exact match?  Was it for speed’s sake that you chose not to do that?
 \(But how much slower could it be\, since you have to get to the GV or
the pad ultimately\, anyway\, to find out the variable name?\)

> @&#8203;\{\[$x\]\} is a special case where the undef value
> unambiguously came from $x; but it fails in the more general case&#8203;:
> 
>     say @&#8203;\{\[$x\, $y\]\}
> 
> here\, using the elements of the array will raise two 'uninit' warnings\,
> each of which may have been derived from either $x or $y\.
> 
> 


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Nov 21 12&#8203;:47&#8203;:59 2011\, sprout wrote&#8203;:
> On Mon Nov 21 09&#8203;:25&#8203;:42 2011\, davem wrote&#8203;:
> > Well\, nothing's entirely clear cut\.
> > 
> > However\, suppose hypothetically that sin\(undef\) returned undef \(rather
> > than 0\.0\); then I would expect the following&#8203;:
> > 
> >     $ perl \-we 'my \($x\,$y\); $y = 1 \+ sin $x'
> >     Use of uninitialized value $x in sin at \-e line 1\.
> >     Use of uninitialized value in addition \(\+\) at \-e line 1\.
> >     $
> > 
> > i\.e\. the second warning \*shouldn't\* mention $x\.
> > So I think your fix for \#72090 was correct\.
> > 
> > As for the OP's case\,  say @&#8203;\{\[ $x \]\}&#8203;: clearly the current behaviour of
> > blaming $" is wrong\, and certainly fixing so that it blames no
variable is
> > correct\. Whether it should be extra smart and blame $x is debatable\, but
> > on balance I think not\. When looking at the SV that triggered the
> > warninng\, my code distinguishes between a specific SV that happens to be
> > \!SvOK\(\)\, and PL\_sv\_undef\. In the former case it can be quite
aggressive in
> > matching that SV up with a particular pad or package variable\, while in
> > the latter case\, it has to be very conservative\, and usually only follow
> > certain type of op back\.
> 
> One of us must be confused\, as\, at the top level \(the first call\)\,
> find\_uninit\_var does not make such a distinction based on the variable
> itself\, but on the op type\. Only if the op is one that can return undef
> for a \*defined\* value\, or if the op has multiple children that are
> possible suspects\, does it look for an exact match\.  If an op has one
> child\, it assumes it can simply dig deeper\.  That’s why I raised the
> question as to whether \#72090 was really a bug\, because it appeared to
> be by design\.
> 
> Based on your ‘I think not’\, why not change the code always to check for
> an exact match?  Was it for speed’s sake that you chose not to do that?
>  \(But how much slower could it be\, since you have to get to the GV or
> the pad ultimately\, anyway\, to find out the variable name?\)
> 
> > @&#8203;\{\[$x\]\} is a special case where the undef value
> > unambiguously came from $x;

It’s actually not hard to make that work\. I think I now understand your
code better than you do\, as it has been a long time\. :\-\)

I think the real question is&#8203;: Is it \*useful\* to mention the source of
the undefinedness\, for the sake of debugging \(which is why this feature
was added to begin with\)?

> > but it fails in the more general case&#8203;:
> > 
> >     say @&#8203;\{\[$x\, $y\]\}
> > 
> > here\, using the elements of the array will raise two 'uninit' warnings\,
> > each of which may have been derived from either $x or $y\.
> > 
> > 
> 
> 


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Nov 21\, 2011 at 12&#8203;:47&#8203;:59PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> One of us must be confused\, as\, at the top level \(the first call\)\,
> find\_uninit\_var does not make such a distinction based on the variable
> itself\, but on the op type\. Only if the op is one that can return undef
> for a \*defined\* value\, or if the op has multiple children that are
> possible suspects\, does it look for an exact match\.  If an op has one
> child\, it assumes it can simply dig deeper\.  That’s why I raised the
> question as to whether \#72090 was really a bug\, because it appeared to
> be by design\.
> 
> Based on your ‘I think not’\, why not change the code always to check for
> an exact match?  Was it for speed’s sake that you chose not to do that?
>  \(But how much slower could it be\, since you have to get to the GV or
> the pad ultimately\, anyway\, to find out the variable name?\)

Well\, it was nearly 8 years ago that I implemented that\, so my memory
is likely to be hazy\!

However\, a quick 'make test' with the top level call to find\_uninit\_var
set to match=1\, gives a slew of failures\, the first one being the
equivalent of&#8203;:

$ perl \-we'my @&#8203;a; $a\[0\] \+ 1'
&nbsp;    before&#8203;: Use of uninitialized value $a\[0\] in addition \(\+\) at \-e line 1\.
&nbsp;    after&#8203;:  Use of uninitialized value in addition \(\+\) at \-e line 1\.

Here\, the $a\[0\] is PL\_sv\_undef\, but its ok to match against PL\_sv\_undef
since there's no other possible source for the undef value; in the case of
$a\[0\] \+ $a\[1\] we can't identify the elements\, which is why we switch to
match=1\.



\-\- 
You're only as old as you look\.

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Nov 21\, 2011 at 12&#8203;:56&#8203;:58PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> It’s actually not hard to make that work\. I think I now understand your
> code better than you do\, as it has been a long time\. :\-\)
> 
> I think the real question is&#8203;: Is it \*useful\* to mention the source of
> the undefinedness\, for the sake of debugging \(which is why this feature
> was added to begin with\)?

Its always worth mentioning the variable name if
\* it's easy\-ish to code;
\* it doesn't give false positives;

\-\- 
In the 70's we wore flares because we didn't know any better\.
What possible excuse does the current generation have?

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Nov 21 16&#8203;:43&#8203;:48 2011\, davem wrote&#8203;:
> On Mon\, Nov 21\, 2011 at 12&#8203;:56&#8203;:58PM \-0800\, Father Chrysostomos via RT
wrote&#8203;:
> > It’s actually not hard to make that work\. I think I now understand your
> > code better than you do\, as it has been a long time\. :\-\)
> > 
> > I think the real question is&#8203;: Is it \*useful\* to mention the source of
> > the undefinedness\, for the sake of debugging \(which is why this feature
> > was added to begin with\)?
> 
> Its always worth mentioning the variable name if
> \* it's easy\-ish to code;
> \* it doesn't give false positives;
> 

In the light of that response\, would you consider \#72090 not\-a\-bug? 
\(The easiest way to fix this one is to revert the change\.\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Nov 21 17&#8203;:55&#8203;:36 2011\, sprout wrote&#8203;:
> On Mon Nov 21 16&#8203;:43&#8203;:48 2011\, davem wrote&#8203;:
> > On Mon\, Nov 21\, 2011 at 12&#8203;:56&#8203;:58PM \-0800\, Father Chrysostomos via RT
> wrote&#8203;:
> > > It’s actually not hard to make that work\. I think I now understand
your
> > > code better than you do\, as it has been a long time\. :\-\)
> > > 
> > > I think the real question is&#8203;: Is it \*useful\* to mention the source of
> > > the undefinedness\, for the sake of debugging \(which is why this
feature
> > > was added to begin with\)?
> > 
> > Its always worth mentioning the variable name if
> > \* it's easy\-ish to code;
> > \* it doesn't give false positives;
> > 
> 
> In the light of that response\, would you consider \#72090 not\-a\-bug? 
> \(The easiest way to fix this one is to revert the change\.\)
> 

I am certainly one to oscillate\.  I’m now beginning to agree with your
first response and Nicholas Clark’s bug report\.  It can be confusing to
see ‘Use of uninitialized value $x in addition’ mentioned for
1\+\(@&#8203;\{\[length $x\]\}\)\[0\]\.  It’s probably best just to leave the variable
name out\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Nov 21\, 2011 at 10&#8203;:06&#8203;:01PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> On Mon Nov 21 17&#8203;:55&#8203;:36 2011\, sprout wrote&#8203;:
> > On Mon Nov 21 16&#8203;:43&#8203;:48 2011\, davem wrote&#8203;:
> > > On Mon\, Nov 21\, 2011 at 12&#8203;:56&#8203;:58PM \-0800\, Father Chrysostomos via RT
> > wrote&#8203;:
> > > > It’s actually not hard to make that work\. I think I now understand
> your
> > > > code better than you do\, as it has been a long time\. :\-\)
> > > > 
> > > > I think the real question is&#8203;: Is it \*useful\* to mention the source of
> > > > the undefinedness\, for the sake of debugging \(which is why this
> feature
> > > > was added to begin with\)?
> > > 
> > > Its always worth mentioning the variable name if
> > > \* it's easy\-ish to code;
> > > \* it doesn't give false positives;
> > > 
> > 
> > In the light of that response\, would you consider \#72090 not\-a\-bug? 
> > \(The easiest way to fix this one is to revert the change\.\)
> > 
> 
> I am certainly one to oscillate\.  I’m now beginning to agree with your
> first response and Nicholas Clark’s bug report\.  It can be confusing to
> see ‘Use of uninitialized value $x in addition’ mentioned for
> 1\+\(@&#8203;\{\[length $x\]\}\)\[0\]\.  It’s probably best just to leave the variable
> name out\.

on balance I agree

\-\- 
My get\-up\-and\-go just got up and went\.

**Comment by p5pRT** [comment]
### From @cpansprout

This is now fixed by 73b42ce\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
